lib/MusicBrainz/Server/Data/Release.pm

```diff
@@ -880,54 +883,155 @@ sub can_merge {
+           AND newt.medium = newm.id
+           AND oldt.medium = oldm.id
+           AND newt.position = oldt.position
+           AND newr.id = newt.recording
+           AND oldr.id = oldt.recording
+           AND newr.id != oldr.id
+         GROUP BY new_medium_position,
+                  new_track_number,
+                  new_track_position,
+                  newr.id
+         ORDER BY new_medium_position,
+                  new_track_position
+    }, $new_release_id, \@old_release_ids);
+
+    state $json = JSON::XS->new->utf8(0);
+    my %target_count;
+
+    for my $possible_merge (@{$possible_merges}) {
+        $possible_merge->{new_recording} = $json->decode($possible_merge->{new_recording});
+        $possible_merge->{old_recordings} = [map { $json->decode($_) } @{$possible_merge->{old_recordings}}];
+
+        # We need to make sure that for each old recording, there is only 1
+        # new recording to merge into. If there is > 1, then it's not clear
+        # what we should merge into.
+        $target_count{$_->{id}}++ for @{$possible_merge->{old_recordings}};
     }
 
-    my %recording_by_position;
-    for my $m_pos (keys %medium_by_position) {
-        # must have at least two mediums
-        my @mediums = @{ $medium_by_position{$m_pos} };
-        next if @mediums <= 1;
-        # all mediums must have the same number of tracks
-        my $track_count = $mediums[0]->track_count;
-        next if grep { $_->track_count != $track_count } @mediums;
-        # group recordings by track position
-        $recording_by_position{$m_pos} = {};
-        for my $medium (@mediums) {
-            for my $tr ($medium->all_tracks) {
-                my $tr_pos = $tr->position;
-                if (exists $recording_by_position{$m_pos}->{$tr_pos}) {
-                    push @{ $recording_by_position{$m_pos}->{$tr_pos} }, $tr->recording;
-                }
-                else {
-                    $recording_by_position{$m_pos}->{$tr_pos} = [ $tr->recording ];
-                }
+    # MBS-8614. Track recording merges, to resolve cases where a recording is
+    # a merge source on one track (after which it gets deleted), and a merge
+    # target on another track (in which case we should instead use the ID of
+    # the target from the first merge). Example where recording 3 should be
+    # merged into recording 2:
+    # 1 -> 2
+    # 3 -> 1
+    my %merge_targets;
+
+    $possible_merges = [grep {
+        # Assigned to avoid confusion with unrelated uses of $_ below.
+        my $possible_merge = $_;
+
+        my ($new_recording, $old_recordings) = @{$possible_merge}{qw(
+            new_recording
+            old_recordings
+        )};
+
+        # As mentioned above, remove old (source) recordings that have
+        # multiple targets.
+        #
+        # Perl fact: this assignment modifies the array in $possible_merge.
+        @{$old_recordings} = grep { $target_count{$_->{id}} == 1 } @{$old_recordings};
```

Comment:
I'm confused by this. Wouldn't we want to just block the release merge if it leads to a situation like this one? What is the alternative? Just leaving the old recording behind, orphaned?

Reply:
Oh, I see that a following commit does exactly that, IIUC. Nevermind then, I guess!

Reply:
Yes, that's just how it always worked since 7eaca53cbc2e75f83237098868cbb6d8182203a8, but I changed it to fail the edit in one of the commits.